# Problemas de prueba de hipótesis

In [8]:
# Se importan las librerías:
from scipy import stats
from scipy.stats import tukey_hsd # Librería utilizada para poder sacar conclusiones a partir del test realizado en el segundo ejercicio
import numpy as np
from statsmodels.stats.multicomp import pairwise_tukeyhsd # Libería utilizada para sacar conclusiones de forma más clara a partir del test realizado en el segundo ejercicio

## Ejercicio 1

Usted es un nutricionista que investiga dos tipos diferentes de dietas para ver si existe una diferencia significativa en la pérdida de peso después de un mes. Eliges dos grupos aleatorios de personas; un grupo sigue la primera dieta y el otro sigue la segunda. Al final del mes, se registra la pérdida de peso (en kg) de cada persona:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

Con estos datos se busca responder a la siguiente pregunta: ¿Existe una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta?

Para sacar conclusiones, siga los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.

-Realizar la prueba para comprobar la hipótesis. Puede utilizar la prueba t de Student.

-Analizar las conclusiones.

**Hipótesis Nula, H0:** No hay ninguna diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y los que siguieron la segunda

**Hipótesis Alternativa, H1:** Hay una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y los que siguieron la segunda

In [4]:
# Datos: 
diet1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5];
diet2 = [3.0, 3.2, 2.9, 2.8, 3.0, 3.2];

# Se establece un Nivel de Significancia del 0.05 para poder establecer un límite a partir del cual dar una afirmación:
alpha = 0.05;


Para poder realizar el **t-Test**, se debe confirmar que los datos son **independientes**, siguen una **distribución normal** y tienen una **varianza estadísticamente similar**. 

En primer lugar, se puede asegurar que los **datos son independientes**, dado que **no están influidos entre sí** ni por ningún **experimento previo**.

A continuación, se van a realizar diferentes comprobaciones, para poder llegar a aplicar el t-Test. Esto implica que, primero, se va a comprobar que los **datos siguen una distribución normal** (lo suficientemente normal, de acuerdo con el nivel de significancia establecido). En caso **afirmativo**, se pasará a comprobar si tienen una **varianza estadísticamente similar** (de nuevo, de acuerdo con el nivel de significancia establecido). Si se pasan **ambas pruebas**, se podrá aplicar el **t-Test**, en caso de **fallar** en alguna **no se aplicará**. 

Se van a ir **imprimiendo mensajes** según se comprueben los distintos pasos para **informar al usuario** del proceso. 

In [ ]:
# Se aplica el SHAPIRO TEST para comprobar si los datos siguen una distribución normal (shapiro se utiliza para grupos de datos pequeños):
# Solo se pasa esta primera prueba si AMBOS grupos siguen una distribución normal
# Se considera que la distribución es normal si el resultado del SHAPIRO TEST es mayor o igual al Nivel de Significancia establecido (alpha = 0.05)
if (stats.shapiro(diet1).pvalue and stats.shapiro(diet2).pvalue) >= alpha:

    print (f" Ambos grupos de datos siguen una distribución normal\n");

    # Se aplica el LEVENE TEST para comprobar si la varianza de los datos es estadísticamente similar:
    # Se consiguera estadísticamente similar si el resultado del LEVENE TEST es mayor o igual al Nivel de Significancia establecido (alpha = 0.05)
    if (stats.levene(diet1, diet2).pvalue) >= alpha:

        print (f" Ambos grupos de datos tienen una varianza estadísticamente similar\n");
        print (f" Por ello se aplica el t-test para los dos grupos de datos específicados...\n");

        # Al haber pasado ambas pruebas, se puede aplicar el t-Test y obtener el P-Value para poder extraer conclusiones de las Hipótesis.
        p_Value = stats.ttest_ind(diet1, diet2).pvalue;

        print (f" P-Value: {p_Value}\n");

    # No se ha pasado el LEVENE TEST (las varianzas no son estadísticamente similares)
    else: print (f"... pero no tienen una varianza estadísticamente similar\n") 

# No se ha pasado el SHAPITO TEST (al menos uno de los grupos no sigue una distribución normal)    
else: print(" Al menos uno de tus grupos de datos no sigue una distribución normal\n");

 Ambos grupos de datos siguen una distribución normal

 Ambos grupos de datos tienen una varianza estadísticamente similar

 Por ello se aplica el t-test para los dos grupos de datos específicados...

 P-Value: 0.008652119393753342



Los grupos de datos han pasado las tres pruebas, por lo que se ha podido aplicar el t-Test y obtener el P-Value. 

El P-Value ha sido muy bajo, lo que implica que se rechaza la Hipótesis Nula, aceptando así la Hipótesis Alternativa: 

<div style = "text-align : center; font-style: italic;"> 
Hay una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y los que siguieron la segunda
</div>

## ANOVA

**ANOVA** (*Analysis of Variance - Análisis de variación*) Es una técnica estadística utilizada para comparar las medidas de dos o más grupos. La idea detrás de ANOVA es descomponer la variabilidad total de los datos en dos componentes: variabilidad entre grupos y variabilidad dentro de los grupos:

- **Variabilidad entre grupos**: Esta variabilidad se refiere a las diferencias entre las medias del grupo. Si esta variabilidad es considerablemente mayor que la variabilidad dentro del grupo, podría ser una indicación de que al menos una de las medias del grupo es diferente.

- **Variabilidad dentro del grupo**: Esta variabilidad se refiere a la dispersión de los datos dentro de cada grupo. Si todos los grupos tienen una variabilidad similar, entonces cualquier diferencia notable en las medias de los grupos podría considerarse significativa.

Las hipótesis en ANOVA suelen incluir:

- **Hipótesis nula** ($H₀$): Las medias de todos los grupos son iguales.

- **Hipótesis alternativa** ($H₁$): Al menos una de las medias del grupo es diferente.

Si el resultado de la prueba ANOVA es significativo (por ejemplo, un valor p inferior a un umbral como 0,05), esto sugiere que al menos la media de un grupo es diferente.

## Ejercicio 2

Un agricultor decide probar tres tipos diferentes de fertilizantes para determinar si alguno es superior en términos de producción de maíz. El agricultor planta maíz en 15 parcelas idénticas y utiliza los tres fertilizantes (5 parcelas para cada tipo). Al final de la temporada mide el rendimiento de maíz (en kg) de cada parcela, con el siguiente resultado:

| Fertilizante 1 | Fertilizante 2 | Fertilizante 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

Con estos datos busca responder la siguiente pregunta: ¿Existe una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes?

Para ayudarte, sigue los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.

-Realizar la prueba ANOVA.

-Analizar las conclusiones.

-Si un fertilizante es mejor que otro, ¿cómo podemos saberlo?

**Hipótesis Nula, H0:** No hay una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes

**Hipótesis Alternativa, H1:** Hay una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes

In [5]:
# Datos:
fertilizer1 = [20, 21, 20, 19, 20];
fertilizer2 = [22, 21, 23, 22, 21];
fertilizer3 = [24, 23, 22, 23, 24];

# Se establece un Nivel de Significancia del 0.05 para poder establecer un límite a partir del cual dar una afirmación:
alpha = 0.05;

Para poder realizar el **ANOVA test**, se debe confirmar que los datos son **independientes**, siguen una **distribución normal** y tienen una **varianza estadísticamente similar**. 

En primer lugar, se puede asegurar que los **datos son independientes**, dado que **no están influidos entre sí** ni por ningún **experimento previo**.

A continuación, se van a realizar diferentes comprobaciones, para poder llegar a aplicar el ANOVA Test. Esto implica que, primero, se va a comprobar que los **datos siguen una distribución normal** (lo suficientemente normal, de acuerdo con el nivel de significancia establecido). En caso **afirmativo**, se pasará a comprobar si tienen una **varianza estadísticamente similar** (de nuevo, de acuerdo con el nivel de significancia establecido). Si se pasan **ambas pruebas**, se podrá aplicar el **ANOVA Test**, en caso de **fallar** en alguna **no se aplicará**. 

Se van a ir **imprimiendo mensajes** según se comprueben los distintos pasos para **informar al usuario** del proceso. 

In [ ]:
# Se aplica el SHAPIRO TEST para comprobar si los datos siguen una distribución normal (shapiro se utiliza para grupos de datos pequeños):
# Solo se pasa esta primera prueba si los tres grupos siguen una distribución normal
# Se considera que la distribución es normal si el resultado del SHAPIRO TEST es mayor o igual al Nivel de Significancia establecido (alpha = 0.05)
if (stats.shapiro(fertilizer1).pvalue and stats.shapiro(fertilizer2).pvalue and stats.shapiro(fertilizer3).pvalue) >= alpha:

    print (f" Tus tres grupos de datos siguen una distribución normal\n");

    # Se aplica el LEVENE TEST para comprobar si la varianza de los datos es estadísticamente similar:
    # Se consiguera estadísticamente similar si el resultado del LEVENE TEST es mayor o igual al Nivel de Significancia establecido (alpha = 0.05)
    if (stats.levene(fertilizer1, fertilizer2, fertilizer3).pvalue) >= alpha:

        print (f" Tus tres grupos de datos tienen una varianza estadísticamente similar\n");
        print (f" Por ello se aplica el ANOVA para los tres grupos de datos específicados...\n");

        # Al haber pasado ambas pruebas, se puede aplicar el ANOVA Test (f_oneway) y obtener el P-Value para poder extraer conclusiones de las Hipótesis.
        p_Value = stats.f_oneway(fertilizer1, fertilizer2, fertilizer3).pvalue;

        print (f" P-Value: {p_Value}\n");

    # No se ha pasado el LEVENE TEST (las varianzas no son estadísticamente similares)
    else: print (f"... pero no tienen una varianza estadísticamente similar\n") 

# No se ha pasado el SHAPITO TEST (al menos uno de los grupos no sigue una distribución normal)    
else: print(" Al menos uno de tus grupos de datos no sigue una distribución normal\n");

 Tus tres grupos de datos siguen una distribución normal

 Tus tres grupos de datos tienen una varianza estadísticamente similar

 Por ello se aplica el ANOVA para los tres grupos de datos específicados...

 P-Value: 0.00014047824793190475



Los grupos de datos han pasado las tres pruebas, por lo que se ha podido aplicar el ANOVA Test y obtener el P-Value. 

El P-Value ha sido muy bajo, lo que implica que se rechaza la Hipótesis Nula, aceptando así la Hipótesis Alternativa: 

<div style = "text-align : center; font-style: italic;"> 
Hay una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes
</div>

Una vez se ha aplicado el ANOVA Test y **se ha confirmado que sí existe una diferencia significativa** en el rendimiento pronedio del maíz, dependiendo del fertilizante empleado, resultaría interesante averiguar cuál de los tres tipos de fertilizante estudiados **es mejor que otro**. 

Para ello se va a utilizar la **Tukey's HSD** (Honest Significant Difference), es decir, se va a aplicar una prueba estadística que **compara las posibles combinaciones de pares de grupos** tras realizar una ANOVA y determina si la **diferencia de medias** entre cada par es **significativa**. 

In [7]:
# Se aplica el Tukey HSD test y se imprime
print (tukey_hsd(fertilizer1, fertilizer2, fertilizer3))

Tukey's HSD Pairwise Group Comparisons (95.0% Confidence Interval)
Comparison  Statistic  p-value  Lower CI  Upper CI
 (0 - 1)     -1.800     0.010    -3.143    -0.457
 (0 - 2)     -3.200     0.000    -4.543    -1.857
 (1 - 0)      1.800     0.010     0.457     3.143
 (1 - 2)     -1.400     0.041    -2.743    -0.057
 (2 - 0)      3.200     0.000     1.857     4.543
 (2 - 1)      1.400     0.041     0.057     2.743



En este caso, el Grupo 0 sería el primer grupo de datos dado como argumento al Tukey HSD test, *fertilizer1*, el Grupo 1, el segundo argumento ... y así sucesivamente. 

Sin embargo, como resulta algo confuso de estudiar de esta forma, se va a aplciar la función parwise_tukeyhsd cuyos argumentos han de ser: 

- Valores a analizar (kg obtenidos en cada fertilizante)
- Etiquetas de cada grupo (numeración del fertilizante)
- Nivel de Significancia (alpha)

In [10]:
# Valores a analizar:
kilograms_info = np.concatenate([fertilizer1, fertilizer2, fertilizer3]);

# Etiquetas de cada grupo:
labels = ['Fertilizer 1'] * len(fertilizer1) + ['Fertilizer 2'] * len(fertilizer2) + ['Fertilizer 3'] * len(fertilizer3);

# Realizar Tukey HSD PairWise
tukeyhsd_pairwise_test = pairwise_tukeyhsd(kilograms_info, labels, alpha);

# Mostrar resultados:
print(tukeyhsd_pairwise_test);

     Multiple Comparison of Means - Tukey HSD, FWER=0.05      
   group1       group2    meandiff p-adj  lower  upper  reject
--------------------------------------------------------------
Fertilizer 1 Fertilizer 2      1.8 0.0099 0.4572 3.1428   True
Fertilizer 1 Fertilizer 3      3.2 0.0001 1.8572 4.5428   True
Fertilizer 2 Fertilizer 3      1.4 0.0409 0.0572 2.7428   True
--------------------------------------------------------------


De esta forma, las comparaciones son mucho más claras y es mucho más sencilo establecer conclusiones. 

La columna *meandiff* da la diferencia promedio en el rendimiento entre los dos grupos de fertilizante comparados. Esto implica que, **si es positiva**, el conjunto de datos en **'Group 2' rinde más que el conjunto de datos en 'Group 1'**. De esta forma: 

- El Fertilizante 2 rinde 1.8 veces más que el Fertilizante 1. 

- El Fertilizante 3 rinde 3.2 veces más que el Fertilizante 1.

- El Fertilizante 3 rinde 1.4 veces más que el Fertilizante 1.

Además, comprobando que **ninguno de los límites inferiores del Intervalo de Confianza contiene el 0**, se puede asegurar lo siguiente: 

<div style = "text-align : center;"> 
    <strong> Ranking Fertilizantes </strong><br> <br> 
    Mejor Fertilizante: Fertilizante 3 <br> <br> 
    Segunda Posición: Fertilizante 2 <br> <br> 
    Tercera Posición: Fertilizante 1
</div>
